### PR TITLE
ci: Update CI to use Pixi

### PR DESCRIPTION
**Pull request by hoxbro**
Changes the developer/CI from Hatch to Pixi. This will align it with our other repos.

Supersede #970

### CODE REVIEW

File: doc/developer_guide.md [hunk 1]
```diff
@@ -1,41 +1,189 @@
-# Developer guide
+# Developer Guide
```
Comment (hoxbro) [inline doc/developer_guide.md:1]: This has been copied from Panel. I should have updated it to match the param, but there could potentially be something I haven't updated correctly.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # This is the configuration for pre-commit, a local framework for managing pre-commit hooks
 #   Check out the docs at: https://pre-commit.com/
 
-default_stages: [commit]
+default_stages: [pre-commit]
```
Comment (hoxbro) [inline .pre-commit-config.yaml:1]: Stop emitting a warning.

File: doc/user_guide/Parameter_Types.ipynb [hunk 3]
```diff
@@ -717,8 +717,8 @@
     "class P(param.Parameterized):\n",
     "    p = param.Path('Parameter_Types.ipynb')\n",
     "    f = param.Filename('Parameter_Types.ipynb')\n",
-    "    d = param.Foldername('lib', search_paths=['/','/usr','/share'])\n",
-    "    o = param.Filename('output.csv', check_exists=False)\n",
+    "    d = param.Foldername('lib', search_paths=['/','/usr','/share'], check_exists=False)\n",
```
Comment (hoxbro) [inline doc/user_guide/Parameter_Types.ipynb:720]: Changes made in this file are so we can run this notebook on Windows, too.

File: pyproject.toml [hunk 4]
```diff
@@ -39,13 +40,7 @@ examples = [
     "pandas",
     "panel",
 ]
-doc = [
```
Comment (hoxbro) [inline pyproject.toml:42]: I have removed the developer-focused optional dependencies `doc` and `lint`.

File: tests/testreactive.py [hunk 5]
```diff
@@ -590,7 +590,7 @@ def gen():
 async def test_reactive_gen_pipe():
     def gen(val):
         yield val+1
-        time.sleep(0.05)
+        time.sleep(0.1)
```
Comment (hoxbro) [inline tests/testreactive.py:593]: Avoid the flaky Windows test; use the same fix as https://github.com/holoviz/param/pull/967.

File: .github/workflows/docs.yaml [hunk 6]
```diff
@@ -2,50 +2,71 @@ name: docs
...
-      - name: graphviz
-        run: sudo apt install graphviz graphviz-dev
-      - name: env setup
-        run: |
-          python -m pip install --upgrade pip
-          python -m pip install hatch
-      - name: build docs
-        run: hatch -v run docs:build
+
+  docs_publish:
```
Comment (maximlt) [inline .github/workflows/docs.yaml:58]: Why is this workflow divided in two jobs?

File: .github/workflows/nightly_lock.yaml
Comment (maximlt) [inline .github/workflows/nightly_lock.yaml:1]: I wonder to which extent this is needed for repositories that a low development activity like Param. I guess it doesn't hurt so much.

File: .github/workflows/test.yaml [hunk 7]
```diff
@@ -5,71 +5,128 @@ on:
...
+      fail-fast: false
+      matrix: ${{ fromJson(needs.setup.outputs.matrix) }}
+    timeout-minutes: 30
+    steps:
+      - uses: holoviz-dev/holoviz_tasks/pixi_install@v0
+        with:
+          environments: ${{ matrix.environment }}
+      - name: Test Unit
+        run: |
+          pixi run -e ${{ matrix.environment }} test-unit $COV
```
Comment (maximlt) [inline .github/workflows/test.yaml:98]: Maybe add a `test-unit-cov` command? This is also useful to run as a developer and not just for the CI.

File: pixi.toml [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+panel = "*"
+
+# =============================================
+# =================== TESTS ===================
+# =============================================
+[feature.test-core.dependencies]
+pytest = "*"
+pytest-asyncio = "*"
+pytest-cov = "*"
+pytest-github-actions-annotate-failures = "*"
```
Comment (maximlt) [inline pixi.toml:65]: Not a big fan of this plugin. It's also definitely not part of the `core` deps needed to tun the tests.

File: scripts/conda/build.sh [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+set -euxo pipefail
+
+PACKAGE="param"
+
+python -m build -s .
+
+VERSION=$(python -c "import $PACKAGE; print($PACKAGE._version.__version__)")
```
Comment (maximlt) [inline scripts/conda/build.sh:9]: Out of curiosity, is there a reason this is not `PACKAGE.__version__`?

Comment (hoxbro) [reply scripts/conda/build.sh:9]: This is to be sure that the automatic version of the file is generated. `param.__version__` is more "complex" with `try/except`.

File: .github/workflows/docs.yaml (hunk 6 above)
Comment (maximlt) [reply .github/workflows/docs.yaml:58]: Ah ok, I've never seen that fail before during a docs build. Do you mind simplifying this workflow into one single job? Let's keep things simple!

File: .github/workflows/build.yaml [hunk 10]
```diff
@@ -2,81 +2,141 @@ name: packages
...
+defaults:
+  run:
+    shell: bash -e {0}
+
+env:
+  PYTHON_VERSION: "3.11"
+  PACKAGE: "param"
 
 jobs:
+  waiting_room:
```
[the remark this answers is not in the document]
Comment (maximlt) [reply .github/workflows/build.yaml:23]: > I have limited it to lead maintainers (and me).

Ok. Where is the least of lead maintainers defined?

> I use it to double-check that the version tag is correct.

Too bad we no longer trust the system to produce the right version tag, but that's fine.

File: .github/workflows/nightly_lock.yaml
Comment (maximlt) [reply .github/workflows/nightly_lock.yaml:1]: > Away to get a snapshot of the environment. This can be used to see what package caused a failure. It could also be used to see how the param environments looked a year ago.

Ah yes I missed that the lock file was pushed to S3 too. Sounds useful, when you know about it ;)

File: .github/workflows/test.yaml (hunk 7 above)
Comment (maximlt) [reply .github/workflows/test.yaml:98]: I liked when you said that you moved as much as possible from the config that was on the CI workflows to the pixi config file (e.g. env vars) which is why I was a little surprised to see that there in `test.yaml`.

> I'm cautious about adding more tasks, which could confuse new contributors.

My experience so far has been to copy/paste the commands displayed in the developer guide. I have not relied on `pixi task list` which I think just gives a simple list of tasks available without their definition, and that's pretty cluttered already anyway.

I'm fine with how things are anyway here.

File: pixi.toml (hunk 8 above)
Comment (maximlt) [reply pixi.toml:65]: > I'm a big fan of it. Reduces the need to look into the logs.

My experience of it so far has been to give me the impression that, basically, every Panel PR is failing as it also displays errors that occur during flaky tests.

---

> The same could be set about pytest-cov. The plugin is a no-op if you are not in a Github action and only have pytest as a dependency.

Reminds me of the discussion we recently had about `hatch-conda-build ` :D `pytest-cov` can be used by a user, I don't think `pytest-github-actions-annotate-failures` can?

But anyway, I'm again fine with how this is.

File: pyproject.toml [hunk 11]
```diff
@@ -111,118 +101,15 @@ include = [
...
-
 [tool.pytest.ini_options]
+addopts = "--color=yes"
 python_files = "test*.py"
 filterwarnings = [
   "error",
 ]
 xfail_strict = "true"
 asyncio_mode = "auto"
+asyncio_default_fixture_loop_scope="funcion"
```
Comment (maximlt) [inline pyproject.toml:112]: Typo:

```suggestion
asyncio_default_fixture_loop_scope="function"
```

**Review by maximlt — COMMENTED** [review]
Except a typo and a simplification I'd like to see in the docs workflow, I'm fine with these changes.

@hoxbro I would strongly encourage you to update the HoloViz operating guide (https://holoviz.org/contributing.html#operating-guide) now that the workflows you created are used in the majority of places and `pyctdev` is no longer used in the main repositories. There are many behaviors and explanations as to why things are how they are that are just in your mind and that would be worth sharing with others, current and future contributors/maintainers. If you have custom tooling that is used to inspect/interact with the system you created, it would also be nice to share it in a HoloViz repo.

File: .github/workflows/build.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply .github/workflows/build.yaml:23]: > PyPi also strongly encourages it ([ref](https://docs.pypi.org/trusted-publishers/using-a-publisher/)).

I'm sorry I don't understand, this page links to using a trusted publisher. Are we doing that?

File: pixi.toml (hunk 8 above)
Comment (maximlt) [reply pixi.toml:65]: > The correct thing to be frustrated about is the flaky test themselves 🙂

Both things are frustrating.

**Review by maximlt — APPROVED** [review]
Thanks @hoxbro for doing this work! I'm going to get this merged now as I'd like the next release to use this new infrastructure.

I don't necessarily agree with the increased complexity some workflows got (e.g. publishing waiting room, docs build in two jobs) but I'm not the one doing all the hard work setting up this infra and maintaining it across HoloViz, so it's alright!!

File: .github/workflows/build.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (hoxbro) [reply .github/workflows/build.yaml:23]: > I'm sorry I don't understand, this page links to using a trusted publisher. Are we doing that?

Just to be sure: 

- The waiting room job itself is to ensure that all the building steps succeed. For example, if the PyPi build step fails, the conda package will _not_ be released. 
- Inside the waiting room is an environment `publish`, which people must be approved before release. This was what I tried to reference in the PyPI ref.
image

[the remark this answers is not in the document]
Comment (maximlt) [reply .github/workflows/build.yaml:23]: Ok thanks I got it.

> The waiting room job itself is to ensure that all the building steps succeed. For example, if the PyPi build step fails, the conda package will not be released.

It sounds like it's more for manually checking the two builds, since otherwise it could be automated.

> Inside the waiting room is an environment publish, which people must be approved before release. This was what I tried to reference in the PyPI ref.

Ok I understand better how you use the environment to add this protection rule. Seems not so related to the trusted publisher thing for which `environment` is optional but `permissions` with `id-token: write` is not.

image

By the way I think you should add Jean-Luc and we probably need to update the list of maintainers.

image

All of that sounds like it would be worth being documented somewhere, it's kind of hard to follow.

[the remark this answers is not in the document]
Comment (hoxbro) [reply .github/workflows/build.yaml:23]: > manually checking the two builds

It is automatically done. If the environment were not there, it would do everything without interaction. 

> All of that sounds like it would be worth being documented somewhere

Yes. I will update the holoviz site (soon).
